pinata/pinata.go

```diff
@@ -0,0 +1,225 @@
+
+  // validate JWT	
+	success,err := p.TestAuthentication()
+	if err != nil {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("problem authenticating: %v", err),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+	if !success {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("authentication failed"),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("file does not exist: %s", path),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	url := "https://api.pinata.cloud/pinning/pinFileToIPFS"
+	method := "POST"
+
+	payload := &bytes.Buffer{} // buffer writer
+	writer := multipart.NewWriter(payload)
+	file, fileOpenError := os.Open(path)
+
+	if fileOpenError != nil {
+		return nil, &errors.BifrostError{
+			Err:       fileOpenError,
+			ErrorCode: errors.ErrFileOperationFailed,
+		}
+	}
+
+	part1, err := writer.CreateFormFile("file", filepath.Base(path))
+	if err != nil {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("file does not exist: %s", path),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	_, errFile1 := io.Copy(part1, file)
+	if errFile1 != nil {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("failed to copy multipart data"),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	defer file.Close()
+
+	 metaData := map[string]string{"name": filename}
+	 pinataMetadata,err := json.Marshal(metaData)
+	 if err!= nil {
+		 return nil, &errors.BifrostError{
+        Err:       fmt.Errorf("failed to marshal metadata"),
+        ErrorCode: errors.ErrBadRequest,
+      }
+	 }
+
+	 pinataOptions,err := json.Marshal(options)
+
+
+	 if err!= nil {
+		 return nil, &errors.BifrostError{
+        Err:       fmt.Errorf("failed to convert metadata to JSON"),
+        ErrorCode: errors.ErrBadRequest,
+      }
+	 }
+
+  _ = writer.WriteField("pinataOptions", string(pinataOptions))
+  _ = writer.WriteField("pinataMetadata", string(pinataMetadata))
+
+	err = writer.Close()
+  if err != nil {
+    return nil, &errors.BifrostError{
+      Err:       fmt.Errorf("failed to close multipart writer"),
+      ErrorCode: errors.ErrBadRequest,
+    }
+  }
+
+	client := &http.Client {
+  }
+  req, err := http.NewRequest(method, url, payload)
+
+  if err != nil {
+    return nil, &errors.BifrostError{
+			Err:       err,
+      ErrorCode: errors.ErrBadRequest,
+		}
+  }
+	 req.Header.Add("Authorization", "Bearer " + p.Config().PinataJWT)
+	 
+
+  req.Header.Set("Content-Type", writer.FormDataContentType())
+  res, err := client.Do(req)
+	fmt.Println(req.Body)
+
+	if err != nil {
+    return nil, &errors.BifrostError{
+      Err:       err,
+      ErrorCode: errors.ErrBadRequest,
+    }
+  }
+  defer res.Body.Close()
+  body, err := io.ReadAll(res.Body)
+
+	 if err != nil {
+    return nil, &errors.BifrostError{
+			Err:       err,
+      ErrorCode: errors.ErrBadRequest,
```

Comment:
This should be a file operation error. Bad request refers to an error that would logically be due to the user passing the wrong data format.